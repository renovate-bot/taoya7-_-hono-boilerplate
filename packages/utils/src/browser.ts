const SSR = typeof window === "undefined";

/**
 * Returns true if the client is running on a Mac.
 */
export function isMac(): boolean {
    if (SSR) {
        return false;
    }
    return window.navigator.platform === "MacIntel";
}

/**
 * Returns true if the client is running on Windows.
 */
export function isWindows(): boolean {
    if (SSR) {
        return false;
    }
    return window.navigator.platform === "Win32";
}


/**
 * Returns true if the client is a touch device.
 */
export function isTouchDevice(): boolean {
    if (SSR) {
        return false;
    }
    return window.matchMedia?.("(hover: none) and (pointer: coarse)")?.matches;
}
