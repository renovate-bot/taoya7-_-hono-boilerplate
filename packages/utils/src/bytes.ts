import { isEmpty } from 'lodash'

const BYTE = 1
const KB = BYTE * 1024
const MB = KB * 1024
const GB = MB * 1024
const TB = GB * 1024
const PB = TB * 1024
const regx = /^(-?(?:\d+)?\.?\d+)(b|kb|mb|gb|tb|pb)$/i

export function parseBytes(value: string | number): number | undefined {
  if (isEmpty(value))
    return

  const str = String(value)
  const matches = str.match(regx)

  if (!matches)
    return

  const num = Number.parseFloat(matches[1])
  const type = matches[2].toLowerCase()

  switch (type) {
    case 'pb':
      return num * PB

    case 'tb':
      return num * TB

    case 'gb':
      return num * GB

    case 'mb':
      return num * MB

    case 'kb':
      return num * KB

    case 'b':
      return num * BYTE
  }
}

export function bytes(size: string): number | undefined {
  return parseBytes(size)
}

export function formatBytes(value: string | number): string {
  const mag = Math.abs(Number(value))
  let unit: string
  let size: number | string

  if (mag >= PB) {
    unit = 'PB'
    size = mag / PB
  }
  else if (mag >= TB) {
    unit = 'TB'
    size = mag / TB
  }
  else if (mag >= GB) {
    unit = 'GB'
    size = mag / GB
  }
  else if (mag >= MB) {
    unit = 'MB'
    size = mag / MB
  }
  else if (mag >= KB) {
    unit = 'KB'
    size = mag / KB
  }
  else {
    unit = 'B'
    size = mag / BYTE
  }

  if (String(size).includes('.'))
    size = size.toFixed(1)

  return size + unit
}


/**
 * 将字节转换为人类可读的字符串以供显示
 *
 * @param bytes filesize in bytes
 * @returns Human readable filesize as a string
 */
export function bytesToHumanReadable(bytes: number | undefined) {
  if (!bytes) {
    return "0 Bytes";
  }

  const out = ("0".repeat((bytes.toString().length * 2) % 3) + bytes).match(
    /.{3}/g
  );

  if (!out || bytes < 1000) {
    return bytes + " Bytes";
  }

  const f = (out[1] ?? "").substring(0, 2);
 
  return `${Number(out[0])}${f === "00" ? "" : `.${f}`} ${"  kMGTPEZY"[out.length]
    }B`;
}