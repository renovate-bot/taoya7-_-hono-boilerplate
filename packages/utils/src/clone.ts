import _clone from 'clone'

export function clone<T>(obj: T): T {
  return _clone(obj)
}
