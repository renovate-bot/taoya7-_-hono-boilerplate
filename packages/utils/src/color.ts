import { isEmpty } from 'lodash'
import { colord, extend } from 'colord'
import namesPlugin from 'colord/plugins/names'
import mixPlugin from 'colord/plugins/mix'

extend([namesPlugin, mixPlugin])
const HEX_REGEX = /^#?([0-9A-F]{3}|[0-9A-F]{4}|[0-9A-F]{6}|[0-9A-F]{8})$/i
const RGBA_REGEX
  = /rgba\((\d{1,3}%?),\s*(\d{1,3}%?),\s*(\d{1,3}%?),\s*(\d*(?:\.\d+)?)\)/i
const RGB_REGEX = /rgb\((\d{1,3}%?),\s*(\d{1,3}%?),\s*(\d{1,3}%?)\)/i

/**
 * 混合颜色 Mix color
 *
 * @param firstColor - First color
 * @param secondColor - Second color
 * @param ratio - The ratio of the second color (0 - 1)
 */
export function mixColor(firstColor: string, secondColor: string, ratio: number) {
  return colord(firstColor).mix(secondColor, ratio).toHex()
}
/**
 * 添加颜色透明度
 *
 * @param color - Color
 * @param alpha - Alpha (0 - 1)
 */
export function addColorAlpha(color: string, alpha: number) {
  return colord(color).alpha(alpha).toRgbString()
}
/**
 * Get rgb of color
 *
 * @param color Color
 */
export function getRgbOfColor(color: string) {
  return colord(color).toRgb()
}

/**
 * Is white color
 *
 * @param color - Color
 */
export function isWhiteColor(color: string) {
  return colord(color).isEqual('#ffffff')
}

export function isHexColor(color: string): boolean {
  return HEX_REGEX.test(color)
}

export function isRgba(color: string): boolean {
  return RGBA_REGEX.test(color)
}

export function isRgb(color: string): boolean {
  return RGB_REGEX.test(color)
}

export function hexToRgb(hex: string): number[] {
  if (isEmpty(hex))
    return []

  if (isRgba(hex)) {
    const [, red, green, blue, alpha] = RGBA_REGEX.exec(hex) as any
    return [Number(red), Number(green), Number(blue), Number(alpha)]
  }

  if (isRgb(hex)) {
    const [, red, green, blue] = RGB_REGEX.exec(hex) as any
    return [Number(red), Number(green), Number(blue), 1]
  }

  if (!isHexColor(hex))
    return []

  hex = hex.replace(/^#/, '')

  if (hex.length === 3)
    hex = hex[0] + hex[0] + hex[1] + hex[1] + hex[2] + hex[2]

  let alpha = 1
  if (hex.length === 8) {
    alpha = hexToFloatAlpha(hex.slice(6, 8))
    hex = hex.slice(0, 6)
  }

  const number = Number.parseInt(hex, 16)
  const red = number >> 16
  const green = (number >> 8) & 255
  const blue = number & 255

  return [red, green, blue, alpha]
}

export function isDarkColor(hex: string): boolean {
  if (!isHexColor(hex))
    return true

  return colorBrightness(hex) < 170
}

export function isLightColor(hex: string): boolean {
  return !isDarkColor(hex)
}

export function colorBrightness(hex: string): number {
  const [red, green, blue, alpha] = hexToRgb(hex)
  return (alpha * (red * 299 + green * 587 + blue * 114)) / 1000
}

export function rgbToHex([r, g, b]: number[]): string {
  return `#${((1 << 24) + (r << 16) + (g << 8) + b).toString(16).slice(1)}`
}

export function alpha(hex: string, alpha: number): string {
  const [red, green, blue] = hexToRgb(hex)
  return `rgba(${red}, ${green}, ${blue}, ${alpha})`
}

export function lighten(hex: string, alpha: number): string {
  return rgbToHex(alphaHexToRgb(hex, 1 - alpha, '#fff'))
}

export function darken(hex: string, alpha: number): string {
  return rgbToHex(alphaHexToRgb(hex, 1 - alpha, '#010101'))
}

export function invert(hex: string): string {
  const [red, green, blue] = hexToRgb(hex)
  return rgbToHex([255 - red, 255 - green, 255 - blue])
}

/**
 * Convert rgba to a transparency-adjusted-hex
 * https://stackoverflow.com/a/15898886
 */
export function alphaHexToRgb(
  hex: string,
  alpha: number,
  backgroundHex: string,
): number[] {
  const hexRgb = hexToRgb(hex)
  const backgroundRgb = hexToRgb(backgroundHex)

  return hexRgb.map((color, index) => {
    return colorRange(
      Math.floor(
        colorRange(color) * alpha
        + colorRange(backgroundRgb[index]) * (1 - alpha),
      ),
    )
  })
}

function colorRange(color: number): number {
  return color < 1 ? 1 : color >= 255 ? 255 : color
}

function hexToFloatAlpha(arg: string) {
  const num = Number.parseInt(arg, 16)
  const max = Number.parseInt('0xff', 16)
  return Number((num / max).toFixed(2))
}
