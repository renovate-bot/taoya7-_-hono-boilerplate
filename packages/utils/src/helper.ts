import { type } from './type'

const whiteSpaceRegx
  // eslint-disable-next-line no-control-regex
  = /^[\s\f\n\r\t\u1680\u180E\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200A\u2028\u2029\u202F\u205F\u3000\uFEFF\x09\x0A\x0B\x0C\x0D\x20\xA0]+$/

const numericRegx = /^[+-]?([0-9]*[.])?[0-9]+$/

export function isBoolean(arg: any): boolean {
  return type(arg) === 'boolean'
}

export function isString(arg: any): boolean {
  return type(arg) === 'string'
}

export function isNumber(arg: any): boolean {
  return type(arg) === 'number'
}

export function isNumeric(arg: any): boolean {
  return numericRegx.test(String(arg))
}

export const isArray = Array.isArray

export function isValidArray(arg: any): boolean {
  return isArray(arg) && arg.length > 0
}

export function isNan(arg: any): boolean {
  return isNumber(arg) && Number.isNaN(arg)
}

export function isSet(arg: any): boolean {
  return type(arg) === 'set'
}

export function isWeakSet(arg: any): boolean {
  return type(arg) === 'weakset'
}

export function isMap(arg: any): boolean {
  return type(arg) === 'map'
}

export function isWeakMap(arg: any): boolean {
  return type(arg) === 'weakmap'
}

export function isSymbol(arg: any): boolean {
  return type(arg) === 'symbol'
}

export function isObject(arg: any): boolean {
  return type(arg) === 'object'
}

export function isDate(arg: any): boolean {
  return type(arg) === 'date'
}

export function isRegExp(arg: any): boolean {
  return type(arg) === 'regexp'
}

export function isError(arg: any): boolean {
  return type(arg) === 'error'
}

export function isFunction(arg: any): boolean {
  return type(arg) === 'function'
}

export function isNull(arg: any): boolean {
  return type(arg) === 'null'
}

export function isUndefined(arg: any): boolean {
  return type(arg) === 'undefined'
}

export function isNil(arg: any): boolean {
  return isNull(arg) || isUndefined(arg)
}

export function isPlainObject(arg: any): boolean {
  if (!isObject(arg))
    return false

  const ctor = arg.constructor
  if (typeof ctor !== 'function')
    return false

  const proto = ctor.prototype
  if (!isObject(proto))
    return false

  return proto.hasOwnProperty('isPrototypeOf')
}

export function isEmpty(arg: any): boolean {
  if (isNil(arg))
    return true

  if (isBoolean(arg))
    return false

  if (isNumber(arg))
    return false

  if (isString(arg))
    return arg.length === 0 || whiteSpaceRegx.test(arg)

  if (isFunction(arg) || isArray(arg))
    return arg.length === 0

  switch (type(arg)) {
    case 'file':
    case 'map':
    case 'weakmap':
    case 'set':
    case 'weakset': {
      return arg.size === 0
    }

    case 'object': {
      for (const key in arg) {
        if (Object.hasOwnProperty.call(arg, key))
          return false
      }

      return true
    }

    default:
      break
  }

  return false
}

export function isValid(arg: any): boolean {
  return !isEmpty(arg)
}

export function isEqual(arg1: any, arg2: any): boolean {
  return String(arg1) === String(arg2)
}

export function isTrue(arg1: any): boolean {
  return arg1 === true || arg1 === 'true' || isEqual(arg1, '1')
}

export function isFalse(arg1: any): boolean {
  return arg1 === false || arg1 === 'false' || isEqual(arg1, '0')
}

export function isBool(arg: any): boolean {
  return isTrue(arg) || isFalse(arg)
}

export function isFormData(arg: any): boolean {
  return type(arg) === 'formdata'
}

export function uniqueArray(arg: any): any[] {
  if (!isValidArray(arg))
    return []

  return Array.from(new Set(arg))
}

export function isURL(arg: any): boolean {
  if (isEmpty(arg))
    return false

  try {
    const url = new URL(arg)
    return isValid(url.hostname)
  }
  catch (err) {
    return false
  }
}

export default {
  isBoolean,
  isString,
  isNumber,
  isNumeric,
  isArray,
  isValidArray,
  isNan,
  isSet,
  isMap,
  isSymbol,
  isObject,
  isDate,
  isRegExp,
  isError,
  isFunction,
  isNull,
  isUndefined,
  isNil,
  isPlainObject,
  isEmpty,
  isValid,
  isEqual,
  isTrue,
  isFalse,
  isBool,
  isURL,
  isFormData,
  uniqueArray,
}
