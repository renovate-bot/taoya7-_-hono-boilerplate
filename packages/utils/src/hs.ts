import { isEmpty } from 'lodash'

const second = 1
const minute = second * 60
const hour = minute * 60
const day = hour * 24
const week = day * 7
const year = day * 365
const regx = /^(-?(?:\d+)?\.?\d+)(s|m|h|d|w|y)$/i

function parse(arg: string): number | undefined {
  if (isEmpty(arg))
    return

  const str = String(arg)
  const matches = str.match(regx)

  if (!matches)
    return

  const num = Number.parseFloat(matches[1])
  const type = matches[2].toLowerCase()

  switch (type) {
    case 'y':
      return num * year

    case 'w':
      return num * week

    case 'd':
      return num * day

    case 'h':
      return num * hour

    case 'm':
      return num * minute

    case 's':
      return num * second
  }
}

export function hs(arg: string): number | undefined {
  const value = parse(arg)

  if (value)
    return Math.round(value)
}

export function ms(arg: string): number | undefined {
  const value = hs(arg)

  if (value)
    return value * 1e3
}
