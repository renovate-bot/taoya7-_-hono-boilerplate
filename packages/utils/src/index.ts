export * from './uuid'
export * from './bytes'
export * from './hs'
export * from './date'
export * from './color'
export * from './mime'
export * from './nanoid'
export * from './qs'
export * from './helper'
export * from './random'
export * from './parse'
export * from './clone'
export * from './slugify'
export * from './styles'
export * from './user-agent'
export * from './crypto'
export * from './timers'
export * from './browser'
export * from './url'
