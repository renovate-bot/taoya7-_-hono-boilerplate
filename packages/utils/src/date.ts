import moment from 'moment'
import type { unitOfTime } from 'moment'
import { hs } from './hs'

export function timestamp(): number {
  return Math.floor(Date.now() / 1e3)
}
export function date(str?: string, format?: string) {
  return moment(str, format)
}
export function unixDate(t: number) {
  return moment.unix(t)
}

/**
 * Check if the date is expired
 * @param start 
 * @param end 
 * @param expire expire time 对比时间的单位
 */
export function isDateExpired(
  start: number,
  end: number,
  expire: string,
): boolean {
  return end - start > hs(expire)!
}

/**
 * 时间Diff
 * @param start
 * @param end
 * @param unit 默认返回天数
 * @returns 默认返回天数
 */
export function unixDiff(
  start: number,
  end: number,
  unit: unitOfTime.Base | unitOfTime.Base = 'day',
): number {
  if (start > end || start < 0 || end < 0)
    return 0

  return unixDate(end).diff(unixDate(start), unit)
}
/**
 * Get the date period
 * @param start start date
 * @param value value
 * @param unit unit
 */
export function datePeriod(
  start: number,
  value = 1,
  unit: unitOfTime.Base = 'month',
): number {
  return unixDate(start)
    .add(value, unit as unknown as any)
    .unix()
}
