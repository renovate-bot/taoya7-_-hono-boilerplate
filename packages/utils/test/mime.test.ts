import { expect, it } from 'vitest'
import { mime } from '../src'

it('image mime type', () => {
  expect(mime('.jpg')).toBe('image/jpeg')
  expect(mime('.mp3')).toBe('audio/mpeg')
})
