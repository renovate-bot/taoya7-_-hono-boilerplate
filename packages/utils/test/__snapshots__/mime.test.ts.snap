// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`common file mime type 1`] = `undefined`;

exports[`common image mime type 1`] = `undefined`;
